File: pkg/cover/cover.go
// Copyright 2015 syzkaller project authors. All rights reserved.
// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.

// Package cover provides types for working with coverage information (arrays of covered PCs).
package cover

type Cover map[uint64]struct{}

func FromRaw(raw []uint64) Cover {
	var c Cover
	c.Merge(raw)
	return c
}

func (cov *Cover) Merge(raw []uint64) {
	c := *cov
	if c == nil {
		c = make(Cover)
		*cov = c
	}
	for _, pc := range raw {
		c[pc] = struct{}{}
	}
}

// Merge merges raw into coverage and returns newly added PCs. Overwrites/mutates raw.
func (cov *Cover) MergeDiff(raw []uint64) []uint64 {
	c := *cov
	if c == nil {
		c = make(Cover)
		*cov = c
	}
	n := 0
	for _, pc := range raw {
		if _, ok := c[pc]; ok {
			continue
		}
		c[pc] = struct{}{}
		raw[n] = pc
		n++
	}
	return raw[:n]
}

func (cov *Cover) Serialize() []uint64 {
	res := make([]uint64, 0, len(*cov))
	for pc := range *cov {
		res = append(res, pc)
	}
	return res
}

File: tools/syz-declextract/testdata/cover.c
// Copyright 2025 syzkaller project authors. All rights reserved.
// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.

#include "include/syscall.h"

#define COVER_IOCTL1 1
#define COVER_IOCTL2 2
#define COVER_IOCTL3 3
#define COVER_IOCTL4 4

static void cover_helper(int cmd) {
	int tmp = 0;
	tmp++;
	switch (cmd) {
	case COVER_IOCTL3:
		break;
	case COVER_IOCTL4:
		tmp++;
		break;
	}
}

SYSCALL_DEFINE1(cover, int cmd) {
	int tmp = 0;
	tmp++;
	switch (cmd) {
	case COVER_IOCTL1:
		break;
	case COVER_IOCTL2:
		break;
	case COVER_IOCTL3:
	case COVER_IOCTL4:
		cover_helper(cmd);
		break;
	}
	return tmp;
}
